constraints.pro

```diff
@@ -131,100 +149,119 @@ gen_enforced_field(WorkspaceCwd, 'description', DescriptionWithoutTrailingPeriod
       DescriptionWithoutTrailingPeriod = Description
   ).
 
-% "keywords" must be the same across all workspace packages
-% (and must be unset for the root).
+% All published packages must have the same set of NPM keywords.
 gen_enforced_field(WorkspaceCwd, 'keywords', ['MetaMask', 'Ethereum']) :-
   \+ workspace_field(WorkspaceCwd, 'private', true).
+% Non-published packages do not have any NPM keywords.
 gen_enforced_field(WorkspaceCwd, 'keywords', null) :-
   workspace_field(WorkspaceCwd, 'private', true).
 
-% "homepage" must match the name of the package (based on the workspace
-% directory name) across all workspace packages (and must be unset for the
-% root).
+% The homepage of a published package must match its name (which is in turn
+% based on its workspace directory name).
 gen_enforced_field(WorkspaceCwd, 'homepage', CorrectHomepageUrl) :-
   \+ workspace_field(WorkspaceCwd, 'private', true),
   workspace_basename(WorkspaceCwd, WorkspaceBasename),
-  atomic_list_concat(['https://github.com/MetaMask/controllers/tree/main/packages/', WorkspaceBasename, '#readme'], CorrectHomepageUrl).
+  workspace_field(WorkspaceCwd, 'repository.url', RepoUrl),
+  repo_name(RepoUrl, RepoName),
+  atomic_list_concat(['https://github.com/MetaMask/', RepoName, '/tree/main/packages/', WorkspaceBasename, '#readme'], CorrectHomepageUrl).
+% Non-published packages do not have a homepage.
 gen_enforced_field(WorkspaceCwd, 'homepage', null) :-
   workspace_field(WorkspaceCwd, 'private', true).
 
-% "repository.type" must be "git" for all packages.
-gen_enforced_field(WorkspaceCwd, 'repository.type', 'git').
-
-% "repository.url" must be "https://github.com/MetaMask/controllers.git" for all
-% packages.
-gen_enforced_field(WorkspaceCwd, 'repository.url', 'https://github.com/MetaMask/controllers.git').
-
-% "bugs.url" must be "https://github.com/MetaMask/controllers/issues" for all
-% workspace packages.
-gen_enforced_field(WorkspaceCwd, 'bugs.url', 'https://github.com/MetaMask/controllers/issues') :-
-  \+ workspace_field(WorkspaceCwd, 'private', true).
-
-% "bugs" must unset for the root.
-gen_enforced_field(WorkspaceCwd, 'bugs', null) :-
+% The bugs URL of a published package must point to the Issues page for the
+% repository.
+gen_enforced_field(WorkspaceCwd, 'bugs.url', CorrectBugsUrl) :-
+  \+ workspace_field(WorkspaceCwd, 'private', true),
+  workspace_field(WorkspaceCwd, 'repository.url', RepoUrl),
+  repo_name(RepoUrl, RepoName),
+  atomic_list_concat(['https://github.com/MetaMask/', RepoName, '/issues'], CorrectBugsUrl).
+% Non-published packages do not have a bugs URL.
+gen_enforced_field(WorkspaceCwd, 'bugs.url', null) :-
```

Comment:
I had set `bugs` to `null` rather than `bugs.url` to prevent an empty object from being left in `package.json`.

This rule as-written would still allow `"bugs": {},`. This is what the auto-fixer gave me when removing bugs from the root.

Reply:
Ah. Good point. Fixed in [169622b](https://github.com/MetaMask/controllers/pull/1032/commits/169622ba8cd26a2a863ccf71578703fd0af9725e).